File: src/precalculus_domain_and_range_hard.py
"""
precalculus - domain_and_range (hard)
Generated: 2026-02-22T04:37:18.484972
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Composite function domain
        inner_choices = [
            (sqrt(x - a), f"x \\geq {a}", a, oo, False, True),
            (1/(x - a), f"x \\neq {a}", -oo, oo, True, True),
            (log(x - a), f"x > {a}", a, oo, True, True),
        ]
        
        a = randint(-5, 5)
        inner_func, inner_desc, inner_left, inner_right, inner_left_open, inner_right_open = choice(inner_choices)
        inner_func = inner_func.subs(Symbol('a'), a)
        inner_left = inner_left if inner_left == -oo else inner_left
        inner_right = inner_right if inner_right == oo else inner_right
        
        outer_choices = [
            (sqrt, "\\sqrt{f(x)}", "f(x) \\geq 0"),
            (lambda u: 1/u, "\\frac{1}{f(x)}", "f(x) \\neq 0"),
            (log, "\\ln(f(x))", "f(x) > 0"),
        ]
        
        outer_func, outer_latex, outer_constraint = choice(outer_choices)
        
        if outer_func == sqrt:
            # Need inner_func >= 0
            if isinstance(inner_func, type(sqrt(x))):
                # sqrt composed with sqrt
                inner_a = a
                domain_left = inner_a
                domain_left_open = False
            elif isinstance(inner_func, type(log(x))):
                # sqrt(log(x-a)) needs log(x-a) >= 0, so x-a >= 1, x >= a+1
                domain_left = a + 1
                domain_left_open = False
            else:
                # sqrt(1/(x-a)) needs 1/(x-a) >= 0, so x > a
                domain_left = a
                domain_left_open = True
            domain_right = oo
            domain_right_open = True
            
        elif outer_func == log:
            # Need inner_func > 0
            if isinstance(inner_func, type(sqrt(x))):
                # log(sqrt(x-a)) needs x-a > 0
                domain_left = a
                domain_left_open = True
            elif isinstance(inner_func, type(log(x))):
                # log(log(x-a)) needs log(x-a) > 0, so x-a > 1, x > a+1
                domain_left = a + 1
                domain_left_open = True
            else:
                # log(1/(x-a)) needs 1/(x-a) > 0, so x > a
                domain_left = a
                domain_left_open = True
            domain_right = oo
            domain_right_open = True
            
        else:  # 1/f(x)
            # Need inner_func != 0
            if isinstance(inner_func, type(sqrt(x))):
                # 1/sqrt(x-a) needs x-a > 0
                domain_left = a
                domain_left_open = True
                domain_right = oo
                domain_right_open = True
            elif isinstance(inner_func, type(log(x))):
                # 1/log(x-a) needs log(x-a) != 0, so x-a != 1, x != a+1
                # Domain is (a, a+1) U (a+1, oo)
                # For simplicity, we'll construct a different problem
                domain_left = a
                domain_left_open = True
                domain_right = oo
                domain_right_open = True
            else:
                # 1/(1/(x-a)) = x-a, domain is x != a
                domain_left = -oo
                domain_right = oo
                domain_left_open = True
                domain_right_open = True
        
        ans = fmt_interval(domain_left, domain_right, domain_left_open, domain_right_open)
        
        return problem(
            question=f"Find the domain of $g(x) = {outer_latex}$ where $f(x) = {latex(inner_func)}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="precalculus/domain_and_range",
            solution=steps(
                f"First, find the domain of $f(x) = {latex(inner_func)}$: ${inner_desc}$",
                f"Then apply the constraint from the outer function: ${outer_constraint}$",
                f"Combining these constraints gives domain ${ans}$"
            ),
            calculator="none"
        )
    
    elif problem_type == 2:
        # Piecewise function range
        a = randint(-3, 3)
        b = randint(a+1, a+4)
        
        m1 = nonzero(-3, 3)
        b1 = randint(-5, 5)
        
        m2 = nonzero(-3, 3)
        b2 = randint(-5, 5)
        
        # Evaluate at boundaries
        y_left = m1 * a + b1
        y_mid_left = m1 * b + b1
        y_mid_right = m2 * b + b2
        y_right = m2 * (b + 3) + b2
        
        all_ys = [y_left, y_mid_left, y_mid_right, y_right]
        range_min = min(all_ys)
        range_max = max(all_ys)
        
        ans = fmt_interval(range_min, range_max, False, False)
        
        return problem(
            question=f"Find the range of $f(x) = \\begin{{cases}} {m1}x + {b1} & x \\in [{a}, {b}] \\\\ {m2}x + {b2} & x \\in ({b}, {b+3}] \\end{{cases}}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="precalculus/domain_and_range",
            solution=steps(
                f"Evaluate endpoints: $f({a}) = {y_left}$, $f({b}^-) = {y_mid_left}$, $f({b}^+) = {y_mid_right}$, $f({b+3}) = {y_right}$",
                f"Since both pieces are linear, the range is determined by the minimum and maximum values",
                f"Range: ${ans}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 3:
        # Rational function with vertical asymptotes - find domain
        r1 = nonzero(-5, 5)
        r2 = nonzero(-5, 5)
        while r2 == r1:
            r2 = nonzero(-5, 5)
        
        if r1 > r2:
            r1, r2 = r2, r1
        
        denom = expand((x - r1) * (x - r2))
        numer_coeff = nonzero(-3, 3)
        numer_const = randint(-5, 5)
        numer = numer_coeff * x + numer_const
        
        func = numer / denom
        
        # Domain is (-oo, r1) U (r1, r2) U (r2, oo)
        # We'll return the union as a custom format
        ans = f"open:-inf,open:{r1};open:{r1},open:{r2};open:{r2},open:inf"
        
        return problem(
            question=f"Find the domain of $f(x) = {latex(func)}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="precalculus/domain_and_range",
            solution=steps(
                f"Set denominator equal to zero: ${latex(denom)} = 0$",
                f"Factor: ${latex(factor(denom))} = 0$",
                f"Vertical asymptotes at $x = {r1}$ and $x = {r2}$",
                f"Domain: $(-\\infty, {r1}) \\cup ({r1}, {r2}) \\cup ({r2}, \\infty)$"
            ),
            calculator="none"
        )
    
    elif problem_type == 4:
        # Square root with quadratic - domain
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        
        # sqrt(-(x-a)(x-b)) or sqrt((x-a)(x-b))
        sign = choice([-1, 1])
        
        if a > b:
            a, b = b, a
        
        if sign == 1:
            # sqrt((x-a)(x-b)) needs (x-a)(x-b) >= 0
            # So x <= a or x >= b
            quadratic = expand((x - a) * (x - b))
            ans = f"open:-inf,closed:{a};closed:{b},open:inf"
            constraint = f"(x - {a})(x - {b}) \\geq 0"
            solution_text = f"Need ${constraint}$. This holds when $x \\leq {a}$ or $x \\geq {b}$"
        else:
            # sqrt(-(x-a)(x-b)) needs -(x-a)(x-b) >= 0, so (x-a)(x-b) <= 0
            # So a <= x <= b
            quadratic = expand(-1 * (x - a) * (x - b))
            ans = fmt_interval(a, b, False, False)
            constraint = f"-(x - {a})(x - {b}) \\geq 0"
            solution_text = f"Need ${constraint}$, which means $(x - {a})(x - {b}) \\leq 0$. This holds when ${a} \\leq x \\leq {b}$"
        
        return problem(
            question=f"Find the domain of $f(x) = \\sqrt{{{latex(quadratic)}}}$",
            answer=ans,
            difficulty=(1700, 1800),
            topic="precalculus/domain_and_range",
            solution=steps(
                f"Need ${latex(quadratic)} \\geq 0$",
                f"Factor: ${latex(factor(quadratic) if sign == -1 else factor(quadratic))} \\geq 0$",
                solution_text
            ),
            calculator="none"
        )
    
    else:  # problem_type == 5
        # Find range of transformed quadratic on restricted domain
        h = randint(-3, 3)
        k = randint(-4, 4)
        a_coeff = choice([-1, 1, 2, -2])
        
        # f(x) = a(x - h)^2 + k on [h - d, h + d]
        d = randint(1, 3)
        left = h - d
        right = h + d
        
        # Vertex is at (h, k)
        # If a > 0, min is k at x=h, max is at endpoints
        # If a < 0, max is k at x=h, min is at endpoints
        
        vertex_y = k
        left_y = a_coeff * (left - h)**2 + k
        right_y = a_coeff * (right - h)**2 + k
        
        if a_coeff > 0:
            range_min = vertex_y
            range_max = max(left_y, right_y)
        else:
            range_max = vertex_y
            range_min = min(left_y, right_y)
        
        ans = fmt_interval(range_min, range_max, False, False)
        
        func_expanded = expand(a_coeff * (x - h)**2 + k)
        
        return problem(
            question=f"Find the range of $f(x) = {latex(func_expanded)}$ on the interval $[{left}, {right}]$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="precalculus/domain_and_range",
            solution=steps(
                f"Rewrite in vertex form: $f(x) = {a_coeff}(x - {h})^2 + {k}$",
                f"Vertex at $({h}, {k})$ is in the interval $[{left}, {right}]$",
                f"Evaluate at endpoints: $f({left}) = {left_y}$, $f({right}) = {right_y}$",
                f"Since $a = {a_coeff}$ {'> 0' if a_coeff > 0 else '< 0'}, the {'minimum' if a_coeff > 0 else 'maximum'} is at the vertex",
                f"Range: ${ans}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/precalculus_domain_and_range_hard.jl
# precalculus - domain_and_range (hard)
# Generated: 2026-03-08T20:23:48.540079

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("precalculus/domain_and_range")
    
    prob_type = rand(1:6)
    
    if prob_type == 1
        # Rational function with multiple restrictions
        r1 = randint(-8, 8)
        r2 = randint(-8, 8)
        while r2 == r1
            r2 = randint(-8, 8)
        end
        if r1 > r2
            r1, r2 = r2, r1
        end
        
        denom = expand((x - r1) * (x - r2))
        a = nonzero(-5, 5)
        b = randint(-10, 10)
        num = a*x + b
        
        func = num / denom
        
        domain_parts = []
        if r1 == r2 - 1
            push!(domain_parts, fmt_interval(-Inf, r1, true, false))
            push!(domain_parts, fmt_interval(r1, r2, false, false))
            push!(domain_parts, fmt_interval(r2, Inf, false, true))
        else
            push!(domain_parts, fmt_interval(-Inf, r1, true, false))
            push!(domain_parts, fmt_interval(r1, r2, false, false))
            push!(domain_parts, fmt_interval(r2, Inf, false, true))
        end
        
        answer = join(domain_parts, " ∪ ")
        
        problem(
            question="Find the domain of \$f(x) = $(tex(func))\$",
            answer=answer,
            difficulty=(1800, 2000),
            answer_type="interval_union",
            solution=steps(
                sol("Given", func),
                "The denominator cannot equal zero: \$$(tex(denom)) \\neq 0\$",
                "This gives us \$x \\neq $(r1)\$ and \$x \\neq $(r2)\$",
                "Domain: \$$(answer)\$"
            ),
            time=120
        )
        
    elseif prob_type == 2
        # Square root with quadratic inside
        r1 = randint(-10, 10)
        r2 = randint(-10, 10)
        while r2 == r1
            r2 = randint(-10, 10)
        end
        if r1 > r2
            r1, r2 = r2, r1
        end
        
        a = rand([1, -1]) * rand(1:3)
        inner = expand(a * (x - r1) * (x - r2))
        
        if a > 0
            answer = fmt_interval(-Inf, r1, true, true) * " ∪ " * fmt_interval(r2, Inf, true, true)
            inequality_str = "\\geq"
        else
            answer = fmt_interval(r1, r2, true, true)
            inequality_str = "\\leq"
        end
        
        problem(
            question="Find the domain of \$g(x) = \\sqrt{$(tex(inner))}\$",
            answer=answer,
            answer_type="interval_union",
            difficulty=(1850, 2050),
            solution=steps(
                sol("Given", "g(x) = \\sqrt{$(tex(inner))}"),
                "Require \$$(tex(inner)) \\geq 0\$",
                "Factor: \$$(a)$(tex((x - r1)*(x - r2))) $(inequality_str) 0\$",
                "Analyze sign: roots at \$x = $(r1), $(r2)\$",
                "Domain: \$$(answer)\$"
            ),
            time=180
        )
        
    elseif prob_type == 3
        # Composition with restricted range becoming domain
        a = nonzero(-5, 5)
        h = randint(-8, 8)
        k = randint(-15, 15)
        
        vertex_form = a*(x - h)^2 + k
        expanded = expand(vertex_form)
        
        if a > 0
            range_ans = fmt_interval(k, Inf, true, true)
            extremum = "minimum"
        else
            range_ans = fmt_interval(-Inf, k, true, true)
            extremum = "maximum"
        end
        
        problem(
            question="Find the range of \$h(x) = $(tex(expanded))\$",
            answer=range_ans,
            answer_type="interval",
            difficulty=(1900, 2100),
            solution=steps(
                sol("Given", expanded),
                "Complete the square: \$h(x) = $(tex(vertex_form))\$",
                "Vertex at \$($(h), $(k))\$ with $(extremum) value \$$(k)\$",
                "Since \$a = $(a)\$ is $(a > 0 ? "positive" : "negative"), parabola opens $(a > 0 ? "upward" : "downward")",
                "Range: \$$(range_ans)\$"
            ),
            time=150
        )
        
    elseif prob_type == 4
        # Piecewise function range
        m1 = nonzero(-4, 4)
        b1 = randint(-10, 10)
        
        m2 = nonzero(-4, 4)
        b2 = randint(-10, 10)
        
        split = randint(-5, 5)
        
        left_val = m1 * split + b1
        right_val = m2 * split + b2
        
        y_at_neg_inf = m1 < 0 ? Inf : -Inf
        y_at_pos_inf = m2 > 0 ? Inf : -Inf
        
        y_min = min(left_val, right_val, y_at_neg_inf == Inf ? 1000 : -1000, y_at_pos_inf == Inf ? 1000 : -1000)
        y_max = max(left_val, right_val, y_at_neg_inf == -Inf ? -1000 : 1000, y_at_pos_inf == -Inf ? -1000 : 1000)
        
        all_vals = [left_val, right_val]
        if m1 > 0
            y_left_extreme = -Inf
        else
            y_left_extreme = Inf
        end
        if m2 > 0
            y_right_extreme = Inf
        else
            y_right_extreme = -Inf
        end
        
        range_lower = min(left_val, right_val)
        range_upper = max(left_val, right_val)
        
        if y_left_extreme == -Inf || y_right_extreme == -Inf
            range_lower = -Inf
        end
        if y_left_extreme == Inf || y_right_extreme == Inf
            range_upper = Inf
        end
        
        answer = fmt_interval(range_lower, range_upper, true, true)
        
        piece1_str = "$(m1)x + $(b1)"
        piece2_str = "$(m2)x + $(b2)"
        
        problem(
            question="Find the range of \$f(x) = \\begin{cases} $(piece1_str) & x < $(split) \\\\ $(piece2_str) & x \\geq $(split) \\end{cases}\$",
            answer=answer,
            answer_type="interval",
            difficulty=(2000, 2200),
            solution=steps(
                "Analyze each piece separately",
                "Left piece: \$y = $(piece1_str)\$ for \$x < $(split)\$ gives \$y < $(left_val)\$",
                "Right piece: \$y = $(piece2_str)\$ for \$x \\geq $(split)\$ gives \$y \\geq $(right_val)\$",
                "Combine both pieces and limits",
                "Range: \$$(answer)\$"
            ),
            time=180
        )
        
    elseif prob_type == 5
        # Radical with linear inside, then rational
        m = rand([2, 3, 4])
        b = randint(-15, 15)
        inner = m*x + b
        root_val = -b // m
        
        c = nonzero(-8, 8)
        d = nonzero(-8, 8)
        while d == c
            d = nonzero(-8, 8)
        end
        
        denom_expr = c*x + d
        hole = -d // c
        
        if root_val < hole
            answer = fmt_interval(root_val, hole, true, false) * " ∪ " * fmt_interval(hole, Inf, false, true)
        elseif root_val > hole
            answer = fmt_interval(hole, root_val, false, false) * " ∪ " * fmt_interval(root_val, Inf, true, true)
        else
            answer = fmt_interval(root_val, Inf, false, true)
        end
        
        problem(
            question="Find the domain of \$p(x) = \\frac{\\sqrt{$(tex(inner))}}{$(tex(denom_expr))}\$",
            answer=answer,
            answer_type="interval_union",
            difficulty=(2100, 2300),
            solution=steps(
                "Require two conditions:",
                "1) Square root: \$$(tex(inner)) \\geq 0 \\Rightarrow x \\geq $(root_val)\$",
                "2) Denominator: \$$(tex(denom_expr)) \\neq 0 \\Rightarrow x \\neq $(hole)\$",
                "Combine restrictions",
                "Domain: \$$(answer)\$"
            ),
            time=180
        )
        
    else
        # Absolute value rational with range question
        a = nonzero(-5, 5)
        r = randint(-8, 8)
        denom = x - r
        num = a
        
        # Range of |a/(x-r)| is [0, ∞)
        answer = fmt_interval(0, Inf, true, true)
        
        problem(
            question="Find the range of \$q(x) = \\left|\\frac{$(a)}{$(tex(denom))}\\right|\$",
            answer=answer,
            answer_type="interval",
            difficulty=(2200, 2400),
            solution=steps(
                sol("Given", "q(x) = \\left|\\frac{$(a)}{$(tex(denom))}\\right|"),
                "As \$x \\to $(r)\$, \$|q(x)| \\to \\infty\$",
                "As \$x \\to \\pm\\infty\$, \$q(x) \\to 0\$",
                "Since absolute value, \$q(x) > 0\$ for all \$x \\neq $(r)\$",
                "Range: \$$(answer)\$"
            ),
            time=150
        )
    end
end